### PR TITLE
os/bluestore: optimize blob usage when doing appends/overwrites

### PR DESCRIPTION
**Pull request by ifed01**
We can reuse blob when appending/overwriting some extent instead of allocating another one.
This reduces reference Onode (4Mb total /4K AU) mem usage from 280K to ~32K (max_blob_size=64K and min_alloc_size = 4K)
Read performance is greatly improved too.

### CODE REVIEW

File: src/os/bluestore/bluestore_types.h [hunk 1]
```diff
@@ -456,6 +459,10 @@ struct bluestore_blob_t {
 
   bluestore_blob_t(uint32_t f = 0) : flags(f) {}
 
+  const PExtentVector get_extents() const {
```
Comment (liewegas) [inline src/os/bluestore/bluestore_types.h]: PExtentVector& ?

**Comment by liewegas** [comment]
rebase?

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -1728,6 +1728,64 @@ bool BlueStore::Blob::put_ref(
...
+      new_blen > blen ? blen - b_offset : length)) {
+          return false;
+    }
+  }
+  if (new_blen > blen) {
+    int64_t overflow = new_blen;
+    overflow  -= target_blob_size;
+    assert(blen <= new_blen - overflow);
+    // Unable to decrease the provided length to fit into max_blob_size
+    if (overflow >= length) {
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: should this be if (overflow > 0) ?  if overflow = new_blen - target_blob_size.  also not following what the assert above is for

File: src/os/bluestore/BlueStore.cc [hunk 3]
```diff
@@ -2768,7 +2823,29 @@ void BlueStore::Onode::flush()
...
-
+// WriteContext
+ 
+/// Checks for writes to the same pextent within a blob
+bool BlueStore::WriteContext::has_conflict(
+  BlobRef b,
+  uint64_t loffs,
+  uint64_t blob_len)
+{
+  loffs = P2ALIGN(loffs, blob_len);
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: is blob_len necessarily a power of 2?

Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: This is allocated blob length and hence it has min_alloc_size granularity. That's supposed to be power of 2 IMO.
Moreover currently it's used from do_write_small only that always allocates single AU. Will add an assert though...

File: src/os/bluestore/BlueStore.cc [hunk 4]
```diff
@@ -8436,6 +8516,39 @@ void BlueStore::_do_write_small(
...
     }
+    uint32_t alloc_len = min_alloc_size;
+    auto offset0 = P2ALIGN(offset, alloc_len);
+    auto llen = b->get_blob().get_logical_length();
+    if (!head_read && !tail_read &&
+        b->try_reuse_blob(min_alloc_size,
+                          max_bsize,
+                          offset0 - bstart,
+                          &alloc_len)) {
+      assert(alloc_len = min_alloc_size); // expecting data always
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: == not =

Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: yep

File: src/os/bluestore/BlueStore.cc (hunk 3 above)
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: min_alloc_size is a power of 2, but the blob length needn't be... if min_alloc_size is 64k the blob could be 196k

Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: got it, fixed

**Comment by ifed01** [comment]
retest this please

**Comment by liewegas** [comment]
/home/jenkins-build/build/workspace/ceph-pull-requests/src/os/bluestore/bluestore_types.cc: In member function ‘void bluestore_blob_t::allocated(uint32_t, uint32_t, const AllocExtentVector&)’:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/os/bluestore/bluestore_types.cc:785:14: error: no match for ‘operator=’ (operand types are ‘__gnu_cxx::__normal_iterator<bluestore_pextent_t*, std::vector<bluestore_pextent_t, mempool::pool_allocator<(mempool::pool_index_t)2u, bluestore_pextent_t> > >’ and ‘void’)
     start_it = extents.insert(start_it,

insert doesn't return an iterator until c++14 iirc?

File: src/os/bluestore/BlueStore.cc [hunk 5]
```diff
@@ -1744,6 +1744,64 @@ bool BlueStore::Blob::put_ref(
...
+    if (!get_blob().is_unallocated(
+      b_offset,
+      new_blen > blen ? blen - b_offset : length)) {
+          return false;
+    }
+  }
+  if (new_blen > blen) {
+    int64_t overflow = new_blen;
+    overflow  -= target_blob_size;
+    assert(blen <= new_blen - overflow);
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: could this be triggered if the target blob size changes (is reduced) and the existing blob is already larger than that?

File: src/os/bluestore/BlueStore.cc [hunk 6]
```diff
@@ -8675,6 +8758,39 @@ void BlueStore::_do_write_small(
       logger->inc(l_bluestore_write_small_deferred);
       return;
     }
+    uint32_t alloc_len = min_alloc_size;
+    auto offset0 = P2ALIGN(offset, alloc_len);
+    auto llen = b->get_blob().get_logical_length();
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: llen can move inside the condition

File: src/os/bluestore/BlueStore.cc [hunk 7]
```diff
@@ -1744,6 +1744,64 @@ bool BlueStore::Blob::put_ref(
...
+  } else {
+    //new data overlaps with the existing blob
+    new_blen = MAX(blen, length + b_offset);
+    if (!get_blob().is_unallocated(
+      b_offset,
+      new_blen > blen ? blen - b_offset : length)) {
+          return false;
+    }
+  }
+  if (new_blen > blen) {
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: If i'm understanding this correctly, the new_blen > blen case can only happen if we take the top branch above (b_offset >= blen).  If we take the second path, that means we're writing into an unallocated AU inside a larger blob.  And our write always falls within one AU (this is _do_write_small), so we can't possibly be extending the blob.  Right?

File: src/os/bluestore/BlueStore.cc [hunk 8]
```diff
@@ -8675,6 +8758,39 @@ void BlueStore::_do_write_small(
...
+                        offset0 - bstart,
+                        &alloc_len)) {
+      assert(alloc_len == min_alloc_size); // expecting data always
+                                          // fit into reused blob
+      // if blob wasn't extended - need to check for pending writes desiring to
+      // reuse the same pextent. The rationale is that during GC two chunks
+      // from garbage blobs(compressed?) can share logical space within the same
+      // AU. That's in turn might be caused by unaligned len in clone_range2.
+      // Hence the second write will fail in an attempt to reuse blob at
+      // do_alloc_write().
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: clone range offsets are now always aligned... is this still necessary?

File: src/os/bluestore/BlueStore.cc (hunk 7 above)
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: Oh, _do_write_big uses it too.

File: src/os/bluestore/BlueStore.cc [hunk 9]
```diff
@@ -8859,16 +9030,31 @@ int BlueStore::_do_alloc_write(
...
       } else {
-	assert(b_off == 0);
-	csum_order = std::min(wctx->csum_order, ctz(l->length()));
+        csum_order = std::min(wctx->csum_order, ctz(l->length()));
+      }
+      // try to align blob with max_blob_size to improve
+      // its reuse ratio, e.g. in case of reverse write
+      uint32_t suggested_boff =
+       (wi.logical_offset - (wi.b_off0 - wi.b_off)) % max_bsize;
+      if((suggested_boff % (1 << csum_order)) == 0 &&
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: whitespace

**Comment by ifed01** [comment]
>insert doesn't return an iterator until c++14 iirc?
Actually that's a C++11 feature. Looks like some older gcc lacks its support though. Works for my gcc version 5.4.0. Will make a workaround...

File: src/os/bluestore/BlueStore.cc (hunk 5 above)
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: yep, will fix

File: src/os/bluestore/BlueStore.cc (hunk 8 above)
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: I'd prefer to leave it as is. The rationales:Firstly do you mena

Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: yep, makes sense!

**Pull request merged**